### PR TITLE
runtime: Adding the correct detection of mediated PCIe devices

### LINKED ISSUE #4212: runtime: Allow MDEV VFIO devices to be attached to a PCIe root port
The current implementation in Kata attaches mediated devices to a PCIe bus, for hotplugging we need to add a PCIe mediated device to a PCIe root port.

The issue is that the runtime will only attach a PCI device to a root port if it identifies it as a PCIe device. Currently, it does it by reading the size of the config space of a specific BDF. "If size > 256 then PCIe device else PCI device" and it is attached to the PCI bus and not a root port. 

The check is done by using the BDF of a PCIe device and since MDEV devices usually "only" have a sysfsdev entry (which is in most cases sufficient)  the check fails and does not recognize a MDEV device as PCIe device and hence it is attached to the PCI bus and not to a PCIe root port. 

The sysfsdev entry looks like the following: 
```
 /sys/devices/pci0000:xY/BDF0/BDF1/.../MDEVBDF/UUID
```
This is the complete hierarchy of PCIe busses, ports and devices attached along the path.

### PR DESCRIPTION
**Pull request by zvonkok**
Fixes #4212

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

### CODE REVIEW

**Comment by zvonkok** [comment]
With this fix I am  able to run a GPU container on top of a vGPU mdev device
```
$ ls -l /sys/bus/pci/devices/0000:de:00.4/e9b1d604-57ee-4d29-9910-a8a613f6b0af
lrwxrwxrwx 1 root root    0 May  6 15:22 driver -> ../../../../../../../../../bus/mdev/drivers/vfio_mdev

lrwxrwxrwx 1 root root    0 May  6 02:07 iommu_group -> ../../../../../../../../../kernel/iommu_groups/183

lrwxrwxrwx 1 root root    0 May  6 15:22 mdev_type -> ../mdev_supported_types/nvidia-692
drwxr-xr-x 2 root root    0 May  6 15:22 nvidia
drwxr-xr-x 2 root root    0 May  6 15:22 power
--w------- 1 root root 4096 May  6 15:22 remove
lrwxrwxrwx 1 root root    0 May  5 03:42 subsystem -> ../../../../../../../../../bus/mdev
-rw-r--r-- 1 root root 4096 May  5 03:42 uevent
```
The vfio entry is 183 for the mdev device: 

```
$ sudo ctr --debug run --runtime "io.containerd.kata.v2"  --device /dev/vfio/183  --rm -t "docker.io/nvidia/cuda:11.6.0-base-ubuntu20.04" cuda nvidia-smi
Fri May  6 22:27:56 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GRID A100D-4C       Off  | 00000000:02:00.0 Off |                    0 |
| N/A   N/A    P0    N/A /  N/A |      0MiB /  4096MiB |      7%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**Comment by liubin** [comment]
Thank you @zvonkok 

At least two places need to be fixed:

- Kata needs a formated patch, including commit messages, you can check the format [here](https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#patch-format).

- The test may need to be fixed:

```
building ./ci/../src/runtime/virtcontainers/device/drivers
--- FAIL: TestGetVFIODetails (0.00s)
panic: runtime error: index out of range [-1] [recovered]
	panic: runtime error: index out of range [-1]

goroutine 20 [running]:
testing.tRunner.func1.2({0x[13](https://github.com/kata-containers/kata-containers/runs/6332417796?check_suite_focus=true#step:4:13)61060, 0xc00001c4c8})
	/Users/runner/hostedtoolcache/go/1.17.9/x64/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/Users/runner/hostedtoolcache/go/1.17.9/x64/src/testing/testing.go:1212 +0x218
panic({0x1361060, 0xc00001c4c8})
	/Users/runner/hostedtoolcache/go/1.17.9/x64/src/runtime/panic.go:1038 +0x2[15](https://github.com/kata-containers/kata-containers/runs/6332417796?check_suite_focus=true#step:4:15)
github.com/kata-containers/kata-containers/src/runtime/virtcontainers/device/drivers.getMediatedBDF(...)
	/Users/runner/work/kata-containers/kata-containers/src/runtime/virtcontainers/device/drivers/vfio.go:237
github.com/kata-containers/kata-containers/src/runtime/virtcontainers/device/drivers.getVFIODetails({0x138af5c, 0x18}, {0x0, 0x0})
	/Users/runner/work/kata-containers/kata-containers/src/runtime/virtcontainers/device/drivers/vfio.go:225 +0x2c5
github.com/kata-containers/kata-containers/src/runtime/virtcontainers/device/drivers.TestGetVFIODetails(0x0)
	/Users/runner/work/kata-containers/kata-containers/src/runtime/virtcontainers/device/drivers/vfio_test.go:32 +0x[17](https://github.com/kata-containers/kata-containers/runs/6332417796?check_suite_focus=true#step:4:17)5
testing.tRunner(0xc0000831e0, 0x13a2ec0)
```

**Comment by zvonkok** [comment]
@liubin updated, missed the component in the commit message, updated the failing test. It works on my machine :) 

Can one of you kickoff the tests, @liubin @jodh-intel Thanks.

**Comment by zvonkok** [comment]
@dgibson PTAL

**Comment by zvonkok** [comment]
@jodh-intel @fidencio Who is responsible for the ppc64 test, they are failing because the infrastructure for testing is not working, not something that is related to the PR. 

```
"http://localhost:10248/healthz": dial tcp 127.0.0.1:10248: connect: connection refused.

	Unfortunately, an error has occurred:
		timed out waiting for the condition
```

**Comment by dgibson** [comment]
Sorry, I'm not familiar with mediated devices, so I can't review this.

File: src/runtime/virtcontainers/device/drivers/vfio_test.go [hunk 1]
```diff
@@ -46,4 +46,5 @@ func TestGetVFIODetails(t *testing.T) {
 			assert.Nil(t, err)
 		}
 	}
+
```
Comment (liubin) [inline src/runtime/virtcontainers/device/drivers/vfio_test.go:49]: Delete this line.

**Comment by liubin** [comment]
Can you TAL @bergwolf @lifupan @jiangliu

**Comment by zvonkok** [comment]
For the vGPU use-case we can use two modus-operandi, time-sliced or MIG-sliced partitions of a whole GPU. 
I have tested both use-cases and they work with the changes in this PR. 

```
$ sudo -E PATH=$PATH make install &&   sudo ctr --debug run --runtime "io.containerd.kata.v2"  --device /dev/vfio/203  --rm -t "docker.io/nvidia/cuda:11.6.0-base-ubuntu20.04" cuda nvidia-smi


+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GRID A100D-1-10C    Off  | 00000000:02:00.0 Off |                   On |
| N/A   N/A    P0    N/A /  N/A |                  N/A |     N/A      Default |
|                               |                      |              Enabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| MIG devices:                                                                |
+------------------+----------------------+-----------+-----------------------+
| GPU  GI  CI  MIG |         Memory-Usage |        Vol|         Shared        |
|      ID  ID  Dev |           BAR1-Usage | SM     Unc| CE  ENC  DEC  OFA  JPG|
|                  |                      |        ECC|                       |
|==================+======================+===========+=======================|
|  No MIG devices found                                                       |
+-----------------------------------------------------------------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```
The `GRID A100D-1-10C` indicates a vGPU (GRID) with a special type of MIG-slice namely A100D-1-10C which translates to a MIG with 1 GPU and 10 GB of memory.

**Comment by zvonkok** [comment]
@liubin @jodh-intel Anything needed here? How to move forward?

**Review by liubin — APPROVED** [review]

**Comment by Jimmy-Xu** [comment]
@zvonkok  
Awesome!
What are the results of testing this PR in GPU passthrough mode?

**Comment by zvonkok** [comment]
@Jimmy-Xu This works as expected with GPU passthrough. A passthrough GPU is a "native PCI" device in Kata so it takes another code path.

**Comment by liubin** [comment]
@zvonkok Do you think the according [documentation](https://github.com/kata-containers/kata-containers/blob/main/docs/use-cases/NVIDIA-GPU-passthrough-and-Kata.md#nvidia-vgpu-mode-with-kata-containers) should be updated too? If yes can you help update it in this or another new PR?

**Comment by zvonkok** [comment]
@liubin This was the plan, there is a TODO on vGPU in that document that I left out until this PR works/lands.

**Review by bergwolf — APPROVED** [review]
Nice work! Thanks a lot!

**Comment by zvonkok** [comment]
@fidencio Which versions of Kata should this PR be backported to?

**Pull request merged**